File: apps/client/src/lib/launch.ts
import { invoke } from "@/lib/tauri";
import { shellQuote } from "@/lib/agentboard";

/** One dev-server config from a checkout's Claude Desktop
 * `.claude/launch.json` (`configurations[]`: name / runtimeExecutable /
 * runtimeArgs / port), with live status stamped by the backend — see
 * `launch_configs` in `crates-tauri/tt-app/src/launch.rs`. */
export type LaunchConfigStatus = {
  name: string;
  runtimeExecutable: string;
  runtimeArgs: string[];
  /** Port the server listens on once up — absent configs can't be probed. */
  port?: number | null;
  /** Something accepts TCP connections on `port` right now (always false
   * without a port — `port` itself tells "unknown" from "stopped"). */
  portListening: boolean;
  /** The app session this config was launched into, while that PTY lives —
   * the "focus that pane instead of launching again" target. */
  sessionId?: string | null;
};

/** A folder's launch.json configs + live status. `Err` carries a message
 * worth showing (e.g. a malformed launch.json); a folder without the file
 * resolves `Ok([])`. */
export const launchConfigs = (dir: string) =>
  invoke<LaunchConfigStatus[]>("launch_configs", { dir });

/** Record "config `name` now runs in session `sessionId`" right after typing
 * its command into that pane — the backend prunes the mapping when the PTY
 * dies, and logs the launch gesture to the event log. */
export const launchRegister = (
  dir: string,
  name: string,
  sessionId: string,
  port: number | null,
  command: string,
) => invoke<void>("launch_register", { dir, name, sessionId, port, command });

/** Tokens that read the same quoted or bare — quoting these anyway would
 * render `pnpm dev` as `'pnpm' 'dev'` in the user's own terminal. */
const SAFE_TOKEN = /^[\w@%+=:,./-]+$/;

function quoteToken(token: string): string {
  return SAFE_TOKEN.test(token) ? token : shellQuote(token);
}

/** The command line a config types into its PTY (no trailing `\r` — the
 * caller appends it): `runtimeExecutable runtimeArgs…`, each token quoted
 * only when the shell would otherwise mangle it. */
export function launchCommand(
  cfg: Pick<LaunchConfigStatus, "runtimeExecutable" | "runtimeArgs">,
): string {
  return [cfg.runtimeExecutable, ...cfg.runtimeArgs].map(quoteToken).join(" ");
}

/** Where a listening config serves. Dev servers in launch.json are local by
 * definition — Claude Desktop previews them on localhost the same way. */
export function devServerUrl(port: number): string {
  return `http://localhost:${port}/`;
}

/** What the row's one action slot offers. Independent of the status dot
 * (which tracks `portListening` alone): a pane we launched stays focusable
 * even after the server inside it was Ctrl-C'd. */
export type LaunchAction = "focus" | "external" | "launch";

export function launchAction(cfg: LaunchConfigStatus): LaunchAction {
  // Our own pane wins: even with the port also listening, the pane is the
  // thing the user can actually look at and act in.
  if (cfg.sessionId) return "focus";
  // Port held by something we didn't launch — offering a second launch
  // would just crash into EADDRINUSE.
  if (cfg.portListening) return "external";
  return "launch";
}

File: crates-tauri/tt-app/src/launch.rs
//! Dev servers from a checkout's Claude Desktop `.claude/launch.json` — the
//! Tauri glue over [`tt_agentboard::launch`]. `launch_configs` lists a
//! folder's configs with live status (port probe + which app session runs
//! each); `launch_register` records "config X now runs in session Y" after
//! the client typed the launch command into that session's PTY. The launch
//! itself is frontend-driven, exactly like starting `claude` in a pane —
//! there is no backend spawn path here (the PTY's shell spawn is already
//! recorded via `tt_exec::record_detached_spawn`), so the register event
//! below is what makes the launch gesture visible in the event log.

use std::collections::HashMap;
use std::path::Path;
use std::sync::Mutex;

use tauri::{AppHandle, Manager, State};

/// Which app session (PTY) each launched dev-server config runs in, keyed by
/// `(folder dir, config name)`. In-memory only: PTYs die with the app, so
/// the mapping does too (the terminal's own no-cross-restart-persistence
/// rule). Pruned against the live-PTY registry on every read, so a closed
/// pane stops claiming its config the moment it's gone.
#[derive(Default)]
pub struct LaunchState {
    running: Mutex<HashMap<(String, String), String>>,
}

/// One `launch.json` config plus what the client needs to render its row:
/// the config itself (flattened — `name`/`runtimeExecutable`/`runtimeArgs`/
/// `port`), whether anything is listening on its port, and the app session
/// it runs in when we launched it ourselves. `sessionId` set → "focus that
/// pane"; unset but `portListening` → running outside the app (don't offer
/// a second launch); neither → launchable.
#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LaunchConfigStatus {
    #[serde(flatten)]
    pub config: tt_agentboard::LaunchConfig,
    /// Something accepts TCP connections on `port` right now. Always false
    /// for a config without a port — the client tells "unknown" from
    /// "stopped" by `port` itself.
    pub port_listening: bool,
    pub session_id: Option<String>,
}

/// A folder's dev-server configs with live status. Async over a blocking
/// task: file IO plus one connect probe per distinct port (instant on
/// loopback for the usual refused/accepted cases, but never on the main
/// thread — sync commands dispatch inline on the GTK thread on Linux).
#[tauri::command]
pub async fn launch_configs(
    app: AppHandle,
    dir: String,
) -> Result<Vec<LaunchConfigStatus>, String> {
    tauri::async_runtime::spawn_blocking(move || launch_configs_blocking(&app, dir))
        .await
        .map_err(|e| format!("launch probe task failed: {e}"))?
}

fn launch_configs_blocking(
    app: &AppHandle,
    dir: String,
) -> Result<Vec<LaunchConfigStatus>, String> {
    let file = tt_agentboard::read_launch_file(Path::new(&dir))
        .map_err(|e| format!("launch.json: {e}"))?;
    let Some(file) = file else {
        return Ok(Vec::new());
    };

    // Resolve each config's owning session while pruning dead panes, then
    // drop the lock before the network probes.
    let live = app.state::<crate::terminal::TermState>().live_ids();
    let launch_state = app.state::<LaunchState>();
    let sessions: Vec<Option<String>> = {
        let mut running = launch_state.running.lock().unwrap();
        running.retain(|_, sid| live.contains(sid));
        file.configurations
            .iter()
            .map(|c| running.get(&(dir.clone(), c.name.clone())).cloned())
            .collect()
    };

    // One probe per distinct port — configs may share one (the blog fixture's
    // "all" covers the same server as "blog").
    let mut probed: HashMap<u16, bool> = HashMap::new();
    Ok(file
        .configurations
        .into_iter()
        .zip(sessions)
        .filter(|(config, _)| config.launchable())
        .map(|(config, session_id)| {
            let port_listening = config
                .port
                .map(|p| *probed.entry(p).or_insert_with(|| tt_agentboard::port_listening(p)))
                .unwrap_or(false);
            LaunchConfigStatus { config, port_listening, session_id }
        })
        .collect())
}

/// Record "config `name` from `dir`'s launch.json now runs in session
/// `session_id`" — called by the client right after typing the launch
/// command into that pane. Also the event-log record of the launch gesture
/// itself (root CLAUDE.md: every user-initiated action is logged): the PTY
/// only sees anonymous keystrokes, so without this event a dev-server
/// launch would be indistinguishable from any other typing.
#[tauri::command]
pub fn launch_register(
    state: State<'_, LaunchState>,
    dir: String,
    name: String,
    session_id: String,
    port: Option<u16>,
    command: String,
) {
    tracing::info!(
        dir = %dir,
        config = %name,
        session_id = %session_id,
        port = ?port,
        command = %command,
        "dev_server.launch"
    );
    state.running.lock().unwrap().insert((dir, name), session_id);
}
